Runtime/Triangulator.cs

```diff
@@ -283,49 +288,485 @@ public void Dispose()
 
         public JobHandle Schedule(JobHandle dependencies = default)
         {
-            return new TriangulationJob
-            {
-                preprocessor = Settings.Preprocessor,
-                validateInput = Settings.ValidateInput,
-                restoreBoundary = Settings.RestoreBoundary,
-                refineMesh = Settings.RefineMesh,
-                sloanMaxIters = Settings.SloanMaxIters,
-                concentricShellsParameter = Settings.ConcentricShellsParameter,
-                refinementThresholdArea = Settings.RefinementThresholds.Area,
-                refinementThresholdAngle = Settings.RefinementThresholds.Angle,
-                input = new()
-                {
-                    Positions = Input.Positions,
-                    ConstraintEdges = Input.ConstraintEdges,
-                    HoleSeeds = Input.HoleSeeds,
-                },
-                output = Output,
+            return TriangulateAsync(Settings, new() {
+                Positions = Input.Positions,
+                ConstraintEdges = Input.ConstraintEdges,
+                HoleSeeds = Input.HoleSeeds,
+            }, new OutputData<float2> {
+                Positions = outputPositions,
+                Triangles = triangles,
+                Status = status,
+            }, dependencies);
+        }
+
+        public static JobHandle TriangulateAsync(TriangulationSettings settings, InputData<float2> input, OutputData<float2> output, JobHandle dependencies = default)
```

Comment:
I'm not particularly fond of this API.
I've been thinking about the following solution to introduce the `int2` feature (this is a sketch)

```csharp
    public class Triangulator
    {
        private readonly Triangulator<float> impl = new();
        public JobHandle Schedule(JobHandle dependencies) => impl.Schedule(dependencies);
    }

    public class Triangulator<T> where T : unmanaged
    {
        internal T[] input;
        public JobHandle Schedule(JobHandle dependencies) => default;
    }
```

We could handle which "utils" implementation should we use for given coords (`float2`, `int2`) in `Triangulator<T>`.
This approach maintains the current API while offering what I believe is a very intuitive approach.
An additional advantage is that switching the default coordinates type from `float2` to `double2` can be accomplished by simply changing the generic parameter in `impl`.

Reply:
I really don't like putting things in a class, because that means that the API cannot be used from burst, which would be very nice. That's why I like the static methods. It also more easily communicates that there's no persistent state.

I don't think your API would be able to easily set the utils type. It would have to be additional generic parameters like `Triangulator<float2, float, FloatCoordinateUtils>`, which looks kinda bad.
Or are you saying that this type should be hidden from users? If so, how would they pass inputs with the desired type?

Reply:
Below, I have attached my extended sketchy implementation for `Triangulator<T>`.
I use extensions, although this can be accomplished without them, albeit requiring casting. I believe I've covered all the cases used in your implementation for  `int2`. It preserves the previous API with minimal changes `class InputData -> class Triangulator<float2>.InputData`, but this could be handled as well.

I made this quickly, probably this can be simplified. Some classes could be moved to `Triangulator<T>` and some classes/strucs can be set as `private`. `Job<T1, T2, T3>` could likely be moved into `Triangulator<T>`, allowing us to remove `<T1>` since `<T>` is already higher in the hierarchy.

```csharp
    namespace Experimental
    {
        public class TestsT
        {
            [Test]
            public void GenericsTest()
            {
                // These throws: missing input
                new Triangulator<float2>(Allocator.TempJob).Schedule(default).Complete();
                new Triangulator<int2>(Allocator.TempJob).Schedule(default).Complete();
                new Triangulator(Allocator.TempJob).Schedule(default).Complete();
            }

            [Test]
            public void TExceptionTest()
            {
                new Triangulator<int>(Allocator.TempJob);
            }

            [Test]
            public void SettingsTest()
            {
                using var t1 = new Triangulator<int2>(Allocator.Persistent) { Settings = { RefineMesh = true } };
                using var t2 = new Triangulator(Allocator.Persistent) { Settings = { RefineMesh = true } };
                UnityEngine.Debug.Log(t1.Settings.RefineMesh);
                UnityEngine.Debug.Log(t2.impl.Settings.RefineMesh);
            }

            [Test]
            public void InputTest()
            {
                using var positionsInt = new NativeArray<int2>(64, Allocator.Persistent);
                using var positionsFloat = new NativeArray<float2>(64, Allocator.Persistent);
                using var t1 = new Triangulator<int2>(Allocator.Persistent)
                {
                    Input = { Positions = positionsInt }
                };
                using var t2 = new Triangulator<float2>(Allocator.Persistent)
                {
                    Input = { Positions = positionsFloat }
                };
                using var t3 = new Triangulator(Allocator.Persistent)
                {
                    Input = { Positions = positionsFloat }
                };

                t1.Run();
                t2.Run();
                t3.Run();
            }
        }

        public class Triangulator : IDisposable
        {
            public class TriangulatorSettings { public bool RefineMesh { get; set; } }
            public TriangulatorSettings Settings { get; } = new();
            public Triangulator<float2>.InputData Input { get; } = new();
            public Triangulator<float2>.OutputData Output => impl.Output;
            internal readonly Triangulator<float2> impl; // Can be private
            public Triangulator(Allocator allocator) => impl = new(allocator) { Settings = Settings, Input = Input };
            public JobHandle Schedule(JobHandle dependencies) => impl.Schedule(dependencies);
            public void Run() => impl.Run();
            public void Dispose() => impl.Dispose();
        }

        public class Triangulator<T> : IDisposable where T : unmanaged
        {
            private static readonly IReadOnlyList<Type> supportedTypes = new[] { typeof(int2), typeof(float2) };
            public class InputData { public NativeArray<T> Positions { get; set; } }
            public struct NativeInputData { public NativeArray<T> Positions; }
            public struct OutputData { public NativeList<T> Positions; }
            public TriangulatorSettings Settings { get; internal set; } = new();
            public InputData Input { get; internal set; } = new();
            public OutputData Output => new() { Positions = positions }; 
            private NativeList<T> positions;

            public Triangulator(Allocator allocator)
            {
                if (!supportedTypes.Contains(typeof(T)))
                {
                    throw new NotSupportedException($"{typeof(T)} is not supported.");
                }

                positions = new NativeList<T>(64, allocator);
            }

            public void Dispose()
            {
                positions.Dispose();
            }
        }

        public static class Extensions
        {
            public static JobHandle Schedule(this Triangulator<float2> triangulator, JobHandle dependencies) =>
                new TriangulationJob<float2, float, FloatCoordinateUtils>(triangulator).Schedule(dependencies);
            public static JobHandle Schedule(this Triangulator<int2> triangulator, JobHandle dependencies) =>
                new TriangulationJob<int2, ulong, IntCoordinateUtils>(triangulator).Schedule(dependencies);
            public static void Run(this Triangulator<float2> triangulator) =>
                new TriangulationJob<float2, float, FloatCoordinateUtils>(triangulator).Run();
            public static void Run(this Triangulator<int2> triangulator) =>
                new TriangulationJob<int2, ulong, IntCoordinateUtils>(triangulator).Run();
        }

        [BurstCompile]
        internal struct TriangulationJob<TCoords, TLengthSq, TUtils> : IJob
            where TCoords : unmanaged
            where TLengthSq : unmanaged, IComparable<TLengthSq>
            where TUtils : unmanaged, IUtils<TCoords, TLengthSq>
        {
            private Triangulator<TCoords>.NativeInputData input;
            private Triangulator<TCoords>.OutputData output;
            private readonly TUtils utils;

            public TriangulationJob(Triangulator<TCoords> triangulator)
            {
                input = new() { Positions = triangulator.Input.Positions };
                output = triangulator.Output;
                utils = default;
            }

            public void Execute()
            {
                utils.Log();
            }
        }

        internal struct FloatCoordinateUtils : IUtils<float2, float> 
        { 
            public void Log() => UnityEngine.Debug.Log($"{default(float2)}");
        }
        internal struct IntCoordinateUtils : IUtils<int2, ulong>
        {
            public void Log() => UnityEngine.Debug.Log($"{default(int2)}");
        }
        internal interface IUtils<T1, T2> where T1 : unmanaged where T2 : unmanaged
        {
            void Log();
        }
    }
```

Reply:
```
Please note that currently Triangulator is fully managed. I have an idea to add NativeTriangulator which can be called from unmanaged code (and Triangulator will be just a wrapper).
I have an idea how to can be easly implemented, but let's take it one step at a time :) First, let's introduce this concept Triangulator<T>, and then we can expand upon it.

I believe having both Triangulator<T> and NativeTriangulator<T> would be beneficial. Most users are likely interested in simply calling triangulator.Run() while advanced users might be interested in utilizing the unmanaged features.
```

I don't get the point of having Triangulator and NativeTriangulator. Why not just have a class-less API?

Reply:
Regarding the `Triangulator<T>` and `NativeTriangulator<T>` API

---
There are several reasons why I want to implement it in this way.

1. **Preserving API Consistency:** One primary motivation for implementing it this way is to maintain as similar an API as possible, thereby avoiding introducing breaking changes at this stage.
2. **High-level and Low-level API Separation:** Another reason is to segregate the high-level and low-level APIs. Many users may not be concerned with the jobs pipeline but still desire to leverage the package's performance. Hence, my plan is to introduce `NativeTriangulator<T>` to facilitate full customization (such as adding a `ReadOnly` view, the ability to pass it as a job field, allocating only within a job, etc.). With this struct, users will be responsible for input/output data allocation, and for example allowing for potentially speeding up post-processing when positions are not provided.
3. **Design:** I aim to introduce this functionality without static methods to maintain a clear distinction between `Triangulator` and `NativeTriangulator`. Static methods, if necessary, will be introduced only through extensions. The package's design should not resemble a utility library but rather provide two distinct classes: `Triangulator` and `NativeTriangulator`.

I trust this clears up any uncertainties.

Best
Andrzej

Reply:
Thanks for the reply.

1. I get that. But given that this library is still quite young, and hasn't got a lot of users. This is the perfect time to do architectural changes. And besides, the previous functionality can easily be polyfilled to keep backwards compatibility (while marking things as obsolete).
2. I mean, it's your library, but I don't see how the class based triangulator is so much easier to use than the static methods that I propose? You could even have a separate overload for accepting e.g. `List<Vector3>` positions, if one would want that. People wouldn't need to know anything about the job system to use the `Triangulate` static method, as it would do everything synchronously. I'd say one has to know more about it using your system, as one has to know the difference between `Run` and `Schedule`, whereas `TriangulateAsync` is more clearly job-related. 
4. Again, it's your library. But I really don't understand how this is a better solution. Right now, the Triangulator is just a wrapper for the input, output and settings. It doesn't contain any other state.


```csharp
// These throws: missing input
                new Triangulator<float2>(Allocator.TempJob).Schedule(default).Complete();
                new Triangulator<int2>(Allocator.TempJob).Schedule(default).Complete();
                new Triangulator(Allocator.TempJob).Schedule(default).Complete();
```
This wouldn't be much harder to accidentally do using a static method, as one would have to deliberately pass uninitialized input:
```csharp
Triangulator.InputData<float2> inputData = default;
Triangulator.TriangulateAsync(settings, inputData, output).Complete();
```

To clarify, this is the api I propose:

```csharp
static class Triangulator {
    public static void Triangulate(TriangulationSettings settings, InputData<float2> input, OutputData<float2> output);
    public static void Triangulate(TriangulationSettings settings, InputData<int2> input, OutputData<int2> output);

    public static JobHandle TriangulateAsync(TriangulationSettings settings, InputData<float2> input, OutputData<float2> output, JobHandle dependencies = default);
    public static JobHandle TriangulateAsync(TriangulationSettings settings, InputData<int2> input, OutputData<int2> output, JobHandle dependencies = default);

    // Optional, for even more simplicity where the user doesn't care about reusing the output data allocation
    public static OutputData<float2> Triangulate(TriangulationSettings settings, InputData<float2> input);
    public static OutputData<int2> Triangulate(TriangulationSettings settings, InputData<int2> input);

    // Optional, for even more simplicity where the user wants to use managed lists like List<Vector2> as input, instead of native arrays
    public static ManagedOutput<Vector2> Triangulate(TriangulationSettings settings, ManagedInput<Vector2> input);
}

[Test]
public void InputTest() {
    using var input = new InputData<float2> {
        // ...
    };
    var settings = new TriangulationSettings {
        // ...
    };
    using var output = Triangulator.Triangulate(settings, input);
}

[Test]
public void InputTestAsync() {
    using var input1 = new InputData<float2> {
        // ...
    };
    using var input2 = new InputData<float2> {
        // ...
    };
    using var input3 = new InputData<float2> {
        // ...
    };
    using var output = new OutputData<float2>(Allocator.Persistent);
    var settings = new TriangulationSettings {
        // ...
    };
    var job1 = Triangulator.TriangulateAsync(settings, input, output);
    var job2 = Triangulator.TriangulateAsync(settings, input, output, job1);
    var job3 = Triangulator.TriangulateAsync(settings, input, output, job2);
    job3.Complete();
}
```

Reply:
Many thanks for such a detailed response!

I understand all your points, but I would like to introduce changes gradually. At this stage, I don't want to change the `Triangulator` API. However, as I mentioned before, I want to introduce `NativeTriangulator`, which I believe will address all the cases you mentioned. The API of `NativeTriangulator` will certainly be inspired by all your comments and ideas.

Of course, depending on the development, I do not rule out marking `Triangulator` as `[Obsolete]` in the future and changing the architecture entirely.

I also think that our discussion related to the API is a matter of preference and style, but both of us have the same goal, and both approaches are equivalent in some way. 😀

By the way, in my approach, one could also implement "managed input."